libazureinit/src/provision/ssh.rs

```diff
@@ -67,20 +68,61 @@ pub(crate) fn get_authorized_keys_path_from_sshd() -> io::Result<Option<PathBuf>
     Ok(None)
 }
 
+pub(crate) fn update_sshd_config() -> io::Result<()> {
+    let sshd_config_path = "/etc/ssh/sshd_config";
+    let temp_sshd_config_path = "/etc/ssh/sshd_config.tmp";
+
+    let mut file_content = String::new();
+    {
+        let mut file = OpenOptions::new().read(true).open(sshd_config_path)?;
+        file.read_to_string(&mut file_content)?;
+    }
+
+    let re =
+        Regex::new(r"(?m)^\s*#?\s*PasswordAuthentication\s+no\s*$").unwrap();
+    let modified_content =
+        re.replace_all(&file_content, "PasswordAuthentication yes");
+
+    if modified_content == file_content {
+        return Ok(());
+    }
+
+    let mut temp_file = OpenOptions::new()
+        .write(true)
+        .create(true)
+        .truncate(true)
+        .open(temp_sshd_config_path)?;
+    temp_file.write_all(modified_content.as_bytes())?;
+    temp_file.set_permissions(fs::Permissions::from_mode(0o644))?;
+
+    fs::rename(temp_sshd_config_path, sshd_config_path)?;
+
+    // Restart the sshd service to apply changes
+    Command::new("systemctl")
```

Comment:
This introduces the assumption that `systemd` is the service manager, which we probably don't want to do. We'll want to add a service manager configuration option and use whatever is configured to reload sshd.

Alternatively, we can cause sshd to reload its configuration by sending it a `SIGHUP` signal (documented in its manual page, see https://www.man7.org/linux/man-pages/man8/sshd.8.html). However, that requires that we know its PID which might be a rather tricky endeavor to do safely across all distributions... cloud-init presumably has decided a way to handle this so maybe learning from them would be best.

Reply:
cloud-init's approach:

https://github.com/canonical/cloud-init/blob/main/cloudinit/config/cc_set_passwords.py#L60

I think we'd be best to avoid it altogether with dependency ordering.

Reply:
Removed this section